src/main/java/edu/ucla/library/avpairtree/verticles/WatcherVerticle.java

```diff
@@ -118,59 +132,80 @@ public void start(final Promise<Void> aPromise) {
+                        waveformIndex = rowSize - 1;
+                    }
+                    headerRow = new String[rowSize];
+
+                    for (int index = 0; index < rowSize; index++) {
+                        // Put the new header fields where they belong
+                        if (accessUrlIndex == index) {
+                            headerRow[index] = CsvItem.IIIF_ACCESS_URL_HEADER;
+                        } else if (waveformIndex == index) {
+                            headerRow[index] = CsvItem.WAVEFORM_HEADER;
+                        } else {
+                            // Copy values from the original header row to the new one
+                            headerRow[index] = originalHeader.getName(index + 1); // Row and Header indices are 1-based
+                        }
                     }
 
-                    headerRow[header.size()] = CsvItem.IIIF_ACCESS_URL_HEADER;
                     writer.writeHeader(headerRow);
                 }
 
-                // Stream through all the rows in our CSV file
-                reader.readRows().stream().forEach(row -> {
-                    final String ark = row.get(CsvItem.ITEM_ARK_HEADER);
-
-                    final int columnCount = row.size();
-                    final String[] columns;
-
-                    // Check to see if this CSV file has an existing "IIIF Access URL" column
-                    if (accessUrlIndex == -1) {
-                        columns = new String[columnCount + 1];
-
-                        for (int index = 0; index < columnCount; index++) {
-                            columns[index] = row.get(index + 1); // Strangely, row is 1-based
-                        }
-
-                        if (aArkMap.containsKey(ark)) {
-                            columns[columnCount] = constructAccessURL(aArkMap.get(ark));
-                        }
-                    } else {
-                        columns = new String[columnCount];
-
-                        for (int index = 0; index < columnCount; index++) {
-                            if (accessUrlIndex != index) {
-                                columns[index] = row.get(index + 1); // Strangely, row is 1-based
-                            } else if (aArkMap.containsKey(ark)) {
-                                columns[index] = constructAccessURL(aArkMap.get(ark));
-                            }
+                // Now, stream through all the non-header rows
+                reader.readRows().stream().forEach(originalRow -> {
+                    final String ark = originalRow.get(CsvItem.ITEM_ARK_HEADER);
+                    final CsvItem csvItem = aCsvItemMap.get(ark);
+                    final String[] row = new String[rowSize];
+
+                    for (int index = 0; index < rowSize; index++) {
+                        if (accessUrlIndex == index) {
+                            row[index] = aCsvItemMap.containsKey(ark) ? constructAccessURL(csvItem) : "";
```

Comment:
Interesting Checkstyle didn't complain about the double empty strings(?)

Reply:
I saw that too and was surprised, but decided not to poke the bear if I hadn't woken it